swap/peer.go

```diff
@@ -0,0 +1,189 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+var SigDoesNotMatch = errors.New("signature does not match")
+var DontOwe = errors.New("no negative balance")
+
+// Peer is a devp2p peer for the Swap protocol
+type Peer struct {
+	*protocols.Peer
+	swap *Swap
+}
+
+func NewPeer(p *protocols.Peer, s *Swap) *Peer {
+	return &Peer{
+		Peer: p,
+		swap: s,
+	}
+}
+
+// handleMsg is for handling messages when receiving messages
+func (sp *Peer) handleMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *ChequeRequestMsg:
+		return sp.handleChequeRequestMsg(ctx, msg)
+
+	case *EmitChequeMsg:
+		return sp.handleEmitChequeMsg(ctx, msg)
+
+	case *ErrorMsg:
+		return sp.handleErrorMsg(ctx, msg)
+
+	case *ConfirmMsg:
+		return sp.handleConfirmMsg(ctx, msg)
+	}
+
+	return nil
+}
+
+// handleChequeRequestMsg runs when a peer receives a `ChequeRequestMsg`
+// It is thus run by the debitor
+// So the debitor needs to:
+//   * check that it indeed owes to the requestor (if not, ignore message)
+//   * check serial number
+//   * check amount
+//   * if all is ok, issue the cheque
+func (sp *Peer) handleChequeRequestMsg(ctx context.Context, msg interface{}) (err error) {
+	// check we have indeed a negative balance with the peer
+	var req *ChequeRequestMsg
+	var ok bool
+	var peerBalance int64
+
+	if req, ok = msg.(*ChequeRequestMsg); !ok {
+		return fmt.Errorf("Unexpected message type: %v", err)
+	}
+
+	peer := req.Peer
+
+	sp.swap.lock.Lock()
+	defer sp.swap.lock.Unlock() //TODO: Do we really want to block so long?
+
+	if peerBalance, ok = sp.swap.balances[peer]; !ok {
+		return fmt.Errorf("No exchanges with peer: %v", peer)
+	}
+	// do we actually owe to the remote peer?
+	if peerBalance >= 0 {
+		// TODO: should we send a message to the peer?
+		return DontOwe
+	}
+
+	// balance is negative, send a cheque
+	// TODO: merge with thresholds; need to check for threshold?
+	// if not, any negative balance will result in a cheque at this point
+
+	var cheque *Cheque
+
+	_ = sp.swap.loadCheque(peer)
+	lastCheque := sp.swap.cheques[peer]
+
+	amount := 0 - peerBalance
+
+	//TODO; need to have ChequeRequestMsg to contain last cheque and compare?
+	// emit cheque, send to peer
+	if lastCheque == nil {
+		cheque = &Cheque{
+			Serial: uint64(0),
+			Amount: uint64(amount),
+		}
+	} else {
+		cheque = &Cheque{
+			Serial: lastCheque.Serial + 1,
+			Amount: lastCheque.Amount + uint64(0-peerBalance),
+		}
+	}
+	cheque.Timeout = defaultCashInDelay
+	cheque.Contract = sp.swap.owner.address
```

Comment:
Should this not be `sp.swap.owner.Contract`? `sp.swap.owner.address` is the signer address  from what I understand.

Reply:
You are probably right - I first thought `Contract` to be the address of the issues of a cheque (like saying "I am issuing this cheque"), but that is what the `Sig` is for I suppose. Then the comment says `// address of chequebook, needed to avoid cross-contract submission` - in that case it should be indeed be `sp.swap.owner.address` I believe.